Handle trailing masked column behavior for nested tensor

mikekgfb opened the pull request:
Summary:
Handle trailing masked column behavior for nested tensor by padding during to_padded, to original tensor size

https://github.com/pytorch/pytorch/issues/97111

Test Plan: sandcastle & github

Differential Revision: D45167874

cpuhrsch reviewed (changes requested):
@drisspg - can you take a look?

cpuhrsch commented:
I think this introduces divergent behavior which might lead to silent correctness issues.

I suspect the issue is rather with the check that enables the use of NestedTensors and we shouldn't modify the behavior just to fix a bug.

cc @albanD @jbschlosser @jisaacso

mikekgfb commented:
Belts and suspenders - yes we’re not doing it, but the updated description gives us the right to do so if we want.   Wdyt?

PS: if we do just one, which should we do?

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________
From: Driss Guessous ***@***.***>
Sent: Friday, April 28, 2023 2:12:18 PM
To: pytorch/pytorch ***@***.***>
Cc: Michael Gschwind ***@***.***>; Author ***@***.***>
Subject: Re: [pytorch/pytorch] Handle trailing masked column behavior for nested tensor (PR #100113)

@drisspg commented on this pull request. In torch/nn/modules/transformer. py: > @@ -175,6 +175,9 @@ class TransformerEncoder(Module): (and convert back on output). This will improve the overall performance of TransformerEncoder when padding
ZjQcmQRYFpfptBannerStart
This Message Is From an External Sender

ZjQcmQRYFpfptBannerEnd

@drisspg commented on this pull request.

________________________________

In torch/nn/modules/transformer.py<https://github.com/pytorch/pytorch/pull/100113#discussion_r1180822963>:

> @@ -175,6 +175,9 @@ class TransformerEncoder(Module):
             (and convert back on output). This will improve the overall performance of
             TransformerEncoder when padding rate is high. Default: ``True`` (enabled).

+            When enable_nested_tensor is set to True, trailing masked columns


Aren't we not doing this though since we are updated the to_padded_tensor call?

—
Reply to this email directly, view it on GitHub<https://github.com/pytorch/pytorch/pull/100113#pullrequestreview-1406609859>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AOT4XHIRZ2XZPBZRZ4S5XA3XDQXDFANCNFSM6AAAAAAXMZ7KTE>.
You are receiving this because you authored the thread.Message ID: ***@***.***>

drisspg commented on test/test_transformers.py:
```diff
@@ -656,6 +656,23 @@ def test_mha_native_args(self, nb_heads, bias):
 
             mha(query=x, key=x, value=x, key_padding_mask=pad_mask)
 
+    def test_kpm_mask_trailing_column_with_nested_tensor(self):
```
Make sure to add a `device` arg to this test and use it in the constructor of the x, and mask accordingly on rebase

mikekgfb replied:
Done.

mikekgfb commented on torch/nn/modules/transformer.py:
```diff
@@ -175,6 +175,9 @@ class TransformerEncoder(Module):
             (and convert back on output). This will improve the overall performance of
             TransformerEncoder when padding rate is high. Default: ``True`` (enabled).
 
+            When enable_nested_tensor is set to True, trailing masked columns
```
The intent was to give us future freedom of action if we find that the other implementation can give us optimization opportunities.

Commit:
Handle trailing masked column behavior for nested tensor (#100113)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/100113

Handle trailing masked column behavior for nested tensor by padding during to_padded, to orignial tensor size

https://github.com/pytorch/pytorch/issues/97111

Test Plan: sandcastle & github

Differential Revision: D45167874

fbshipit-source-id: 42495244799492c58ced15116c891ba6f4e6bc02

mikekgfb commented:
@pytorchbot merge

cpuhrsch commented on test/test_transformers.py:
```diff
@@ -688,6 +688,25 @@ def test_mha_native_args(self, nb_heads, bias):
+            norm_first=False,
+            batch_first=False,
+        )
+        transformer_encoder = nn.TransformerEncoder(encoder_layer, num_layers=3, enable_nested_tensor=True).to(device)
```
Is the output of

```
transformer_encoder_with_nt = nn.TransformerEncoder(encoder_layer, num_layers=3, enable_nested_tensor=True).to(device)
output0 = transformer_encoder(src=x, src_key_padding_mask=mask)
```

going to be the same as without nt?

```
transformer_encoder_without_nt = nn.TransformerEncoder(encoder_layer, num_layers=3, enable_nested_tensor=False).to(device)
output1 = transformer_encoder(src=x, src_key_padding_mask=mask)
```

Are `output0` and `output1` going to produce the same output?

cpuhrsch reviewed (changes requested):
I cancelled the merge. I want to make sure that nn.TransformerEncoder behaves the same for both `enable_nested_tensor=True` and `enable_nested_tensor=False`.

cpuhrsch reviewed (approved):
Accepting after further offline discussion. This actually closes the gap in divergent behavior that I had asked about.

cpuhrsch commented:
@pytorchbot merge
